lib/Receiver.js

```diff
@@ -116,53 +119,55 @@ export class Receiver
 				
 			}
 		}
-		
+
 		//If not found yet
 		if (!payload)
 		{
 			//Get last key
-			let key = this.keyring[this.keyring.length-1];
-			
-			//Try ractchet last key
-			for (let i=this.numKeyRatchets; i<MaxRachetAttemtps; ++i)
+			const key = this.keyring[this.keyring.length-1];
+			//Try ratchet last key
+			for (let i = 0; i < MaxRatchetAttempts; ++i)
 			{
-				//Rachet key
-				key = await key.ratchet();
-				
-				//Append to the keyring
-				this.keyring.push(key);
-				
+				//Ratchet key
+				const ratchetKey = await key.ratchet(i);
 				try {
 					//Try to decrypt payload
-					[payload, authTag] = await key.decrypt(type, header, encryptedFrame, extrabytes, skip);
-					//Activate
-					this.schedulePreviousKeysTimeout(key);
+					[payload, authTag] = await ratchetKey.decrypt(type, header, encryptedFrame, extrabytes, skip);
+					// Set working ratchet as new key
+					this.keyring.push(ratchetKey);
+					this.schedulePreviousKeysTimeout(ratchetKey);
```

Comment:
With this approach, if ratcheting passes decryption, do all previous ratchets skill get deleted?

From https://datatracker.ietf.org/doc/html/draft-omara-sframe-00#section-4.3.5.1:

![image](https://user-images.githubusercontent.com/49660118/169403064-3012e623-1cdb-44a1-abdc-cc65a0f6359f.png)

Reply:
@v-kpheng yes. Basically since those ratchets are contained within the `key.rachets` array of the _current_ key. When we delete that key, those ratchets will also stop being used.